src/Microsoft.Extensions.WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,391 @@
+            if (payloadLen > 0)
+            {
+                buffer = await _inbound.ReadAtLeastAsync(payloadLen, cancellationToken);
+
+                if(masked)
+                {
+                    // Need to skip the masking key, since we didn't consume it.
+                    buffer = buffer.Slice(4);
+                }
+                payload = buffer.Slice(0, payloadLen);
+
+                // We don't slice the buffer because we actually want to keep the payload marked unconsumed until the next read
+                // Instead, we capture the position after the payload as the "lastFrameCursor"
+                _lastPayloadEnd = buffer.Slice(payloadLen).Start;
+
+                if (masked)
+                {
+                    unsafe
+                    {
+                        // TODO: Unmask in place? Can we do that safely? If so we can avoid a heap allocation here.
+                        var unmasked = new byte[payload.Length];
+                        payload.CopyTo(unmasked);
+
+                        // Get the masking key into a Span<byte> as efficiently as possible
+                        var maskingKeySpan = Span<byte>.Empty;
+
+                        // Stack space for storing the masking key if needed.
+                        var maskingKeyStorage = stackalloc byte[4];
+
+                        if (maskingKey.IsSingleSpan)
+                        {
+                            // It's already a single span!
+                            maskingKeySpan = maskingKey.First.Span;
+                        }
+                        else
+                        {
+                            // We need to copy it to a single span.
+                            maskingKeySpan = new Span<byte>(maskingKeyStorage, 4);
+                            maskingKey.CopyTo(maskingKeySpan);
+                        }
+
+                        // Either we initialized it with a length, or it came from the maskingKey buffer which should already have been sliced.
+                        Debug.Assert(maskingKeySpan.Length == 4);
+
+                        // Unmask
+                        for (int i = 0; i < payload.Length; i++)
+                        {
+                            unmasked[i] ^= maskingKeySpan[i % 4];
+                        }
+                        payload = ReadableBuffer.Create(unmasked, 0, unmasked.Length);
+                    }
+
+                    // Mark the masking key consumed
+                    _inbound.Advance(buffer.Start);
+                }
+            }
+            else
+            {
+                // If there was a masking key, mark it consumed
+                if (masked)
+                {
+                    _inbound.Advance(buffer.Start);
+                }
+
+                // Clear the "lastFrameCursor" so we don't try to advance anything next receive
+                _lastPayloadEnd = null;
+            }
+
+            return new WebSocketFrame(fin, opcode, payload);
```

Comment:
You should only really expose text and binary frames; the lib should respond to ping automatically without disturbing the caller; close frames should be processed to initiate shutdown, not returned; likewise, the lib should send close frames when shutting down, if it isn't already closed.